[SWDEV-563823][Compiler-rt][ASan] Simplify API Logic 'asan_hsa_amd_ipc_memory_create'.

ampandey-AMD opened the pull request:
- Use reinterpret_cast<uptr> for pointer arithmetic.
 - Add sanitizer interception logic for api 'hsa_amd_pointer_info'.
 - Allow only valid values of ptr and len in non-ASan mode.
    - ptr == Actual agentBaseAddress && len ==
      original_len_used_in_alloc
 - Allow only valid values of ptr and len in ASan mode.
    - Here pinfo is retrieved from external hsa_amd_pointer_info(not
      internal device allocator function AmdgpuMemFuncs::GetPointerInfo)
      - ptr == pinfo.agentBaseAddress && len == pinfo.sizeInBytes
    - ptr == original_ptr_returned_by_ASAN && len ==
      original_len_used_in_alloc

ampandey-AMD committed:
[SWDEV-563823][Compiler-rt][ASan] Simplify API Logic
'asan_hsa_amd_ipc_memory_create'.

 - Use reinterpret_cast<uptr> for pointer arithmetic.
 - Add sanitizer interception logic for api 'hsa_amd_pointer_info'.
 - Allow only valid values of ptr and len in non-ASan mode.
    - ptr == Actual agentBaseAddress && len ==
      original_len_used_in_alloc
 - Allow only valid values of ptr and len in ASan mode.
    - Here pinfo is retrieved from external hsa_amd_pointer_info(not
      internal device allocator function AmdgpuMemFuncs::GetPointerInfo)
      - ptr == pinfo.agentBaseAddress && len == pinfo.sizeInBytes
    - ptr == original_ptr_returned_by_ASAN && len ==
      original_len_used_in_alloc

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2637

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2638

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2639

b-sumner commented:
Does the test application in the ticket print the same "base" "len" and "end" as the uninstrumented version (taking into account address randomization)?

ampandey-AMD commented:
> Does the test application in the ticket print the same "base" "len" and "end" as the uninstrumented version (taking into account address randomization)?

Yes. If we include the current fix with/without intercepting hsa_amd_pointer_info.

### Non-Instrumented O/P
```
hsa_amd_pointer_info: test_buff 0x7c947a815000 base 0x7c947a814000 len 16384 end 0x7c947a818000
hsa_amd_ipc_memory_create: handle: [7a814000][7c94][395301][0][4][0][0][5325e42b]
```
### Instrumented O/P
```
hsa_amd_pointer_info: test_buff 0x7ef7c9509000 base 0x7ef7c9508000 len 16384 end 0x7ef7c950c000
hsa_amd_ipc_memory_create: handle: [c9508000][7ef7][394f4c][0][4][0][0][3a761d78]
```

b-sumner commented:
But the output in the ticket, when -fsanitize=address is not used is:

hsa_amd_pointer_info: test_buff 0x7f3df2400000 base 0x7f3df2400000 len 1024 end 0x7f3df2400400 

Here the buffer and the base are the same, and len is 1024.  Shouldn't our intercepted version of hsa_amd_pointer_info return the same buffer and base, and a len of 1024 instead of 16384?

ampandey-AMD commented:
> But the output in the ticket, when -fsanitize=address is not used is:
> 
> hsa_amd_pointer_info: test_buff 0x7f3df2400000 base 0x7f3df2400000 len 1024 end 0x7f3df2400400
> 
> Here the buffer and the base are the same, and len is 1024. Shouldn't our intercepted version of hsa_amd_pointer_info return the same buffer and base, and a len of 1024 instead of 16384?

Hi @b-sumner , Ok in my previous o/p, the problem was because of LD_LIBRARY_PATH set to asan instrumented libraries and I compiled non-instrumented o/p without flags `-fsanitize=address -shared-libsan`. So my executable was actually linked to asan instrumented libraries of rocr and hipamd.

If I avoid adding the `/opt/rocm/llvm/lib/asan` or `/opt/rocm/lib/asan` from my **LD_LIBRARY_PATH**, then I am getting this non-instrumented O/P

```
hsa_amd_pointer_info: test_buff 0x7b6a28800000 base 0x7b6a28800000 len 1024 end 0x7b6a28800400
hsa_amd_ipc_memory_create: handle: [28800000][7b6a][1feb9][0][4][0][80000000][3d9ec7b6]
```

I think instrumentation here is not kind of problem, we enable the asan interception of those hsa_amd* api calls in the asanified libraries only as our code is bounded by preprocessor macro `SANITIZER_AMDGPU`. We actually don't try to instrument the hsa_api* calls itself.

ampandey-AMD commented:
Hi @b-sumner,
So I tested locally by removing the asan interception code of `hsa_amd_pointer_info`.

Here are the results without `asan_hsa_amd_pointer_info` implemented.

### Instrumented (ASan enabled via -fsanitize=address -shared-libsan with LD_LIBRARY_PATH set to path/to/lib/asan)
```
hsa_amd_pointer_info: test_buff 0x75863cea5000 base 0x75863cea4000 len 16384 end 0x75863cea8000
hsa_amd_ipc_memory_create: handle: [3cea4000][7586][22d9b][0][4][0][0][51fdf52f]
```
### Non-Instrumented (ASan disabled with LD_LIBRARY_PATH not set to /path/to/lib/asan)
```
hsa_amd_pointer_info: test_buff 0x74bd90c00000 base 0x74bd90c00000 len 1024 end 0x74bd90c00400
hsa_amd_ipc_memory_create: handle: [90c00000][74bd][22cd4][0][4][0][80000000][3a14da89]

```

b-sumner commented:
Thanks!  My thinking is that in order to provide acceptable information to our interception of hsa_amd_ipc_memory_create, our interception of hsa_amd_pointer_info should return the same base, len, and end as observed when the application is not built with instrumentation (or running instrumented libraries).  But I would like to hear @bing-ma's take on this.

ampandey-AMD committed:
Address comments of @bing-ma.

Improve logic of asan_hsa_amd_ipc_memory_create based on observations.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2664

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2665

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2666

ampandey-AMD committed:
Improve ASan implementations of hsa api calls based on @bing-ma
suggestions.

This commit again adds the interception of 'hsa_amd_pointer_info' call.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2678

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2679

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2680

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1540,5 +1547,28 @@ hsa_status_t asan_hsa_amd_vmem_address_free(void* ptr, size_t size,
+                                       void* (*alloc)(size_t),
+                                       uint32_t* num_agents_accessible,
+                                       hsa_agent_t** accessible) {
+  void* p = get_allocator().GetBlockBegin(ptr);
```
I think ptr_ is a better name for this variable considering the ipc function implementation.  Consistent naming will help with manitenance.

ampandey-AMD replied:
Done.

ampandey-AMD committed:
Address @b-sumner comments.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2709

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2710

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1540,5 +1547,28 @@ hsa_status_t asan_hsa_amd_vmem_address_free(void* ptr, size_t size,
+                                       uint32_t* num_agents_accessible,
+                                       hsa_agent_t** accessible) {
+  void* ptr_ = get_allocator().GetBlockBegin(ptr);
+  AsanChunk* m = instance.GetAsanChunkByAddr(reinterpret_cast<uptr>(ptr_));
```
Do we not need to guard against null ptr_ here like we do in ipc create?

ampandey-AMD replied:
If `ptr_ == nullptr` then `if` condition at line number 1559 will fail. 

Then `if` condition at line number **1562** for first condition `status == HSA_STATUS_SUCCESS` would fail ambiguously as status is not getting set up with default value. I think we need to initialize this with `status=HSA_STATUS_ERROR_NOT_INITIALIZED`. Is that ok @b-sumner && @bing-ma?

ampandey-AMD replied:
I default initialized status with value `HSA_STATUS_ERROR_NOT_INITIALIZED` in 00898ca25fba0c25f3f1320f7cde6ab07d49434b.

b-sumner replied:
> If `ptr_ == nullptr` then `if` condition at line number 1559 will fail.
> 
But is it OK to call GetAsanChunkByAddr with a null ptr_?  Line 1456 avoids doing this.

ampandey-AMD replied:
Done

ampandey-AMD committed:
Improve code for edge cases testing.

  - Changes to 'asan_hsa_amd_pointer_info'
    - Initialize status with default value
      'HSA_STATUS_ERROR_NOT_INITIALIZE'.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2737

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -52,18 +52,18 @@ static u32 RZSize2Log(u32 rz_size) {
 }
 
-static AsanAllocator &get_allocator();
+static AsanAllocator& get_allocator();
```
I don't think we should change the style of the upstream code.

ampandey-AMD replied:
Done.

ampandey-AMD committed:
Revert "Improve code for edge cases testing."

This reverts commit 00898ca25fba0c25f3f1320f7cde6ab07d49434b.

ampandey-AMD committed:
Improve code for edge cases testing.

- Changes to 'asan_hsa_amd_pointer_info'
  - Initialize status with default value
    'HSA_STATUS_ERROR_NOT_INITIALIZE'.
  - Assure by adding ternary condition to check if ptr_ is not nullptr
    when calling GetASanChunkByAddr.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2804

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2805

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1452,18 +1455,22 @@ static struct AP64<LocalAddressSpaceView> AP_;
+    uptr p = reinterpret_cast<uptr>(ptr);
+    uptr p_ = reinterpret_cast<uptr>(ptr_);
+    if (p == p_ + kPageSize_ && len == m->UsedSize()) {
+      size_t len_ = get_allocator().GetActuallyAllocatedSize(ptr);
```
I think ptr_ might be a better argument here instead of ptr.

ampandey-AMD replied:
Done.

ampandey-AMD committed:
Address comments of @b-sumner.

 Fix Typo.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2894

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2895

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1540,5 +1547,31 @@ hsa_status_t asan_hsa_amd_vmem_address_free(void* ptr, size_t size,
+                     : nullptr;
+  hsa_status_t status = HSA_STATUS_ERROR_NOT_INITIALIZED;
+  if (ptr_ && m) {
+    status = REAL(hsa_amd_pointer_info)(ptr, info, alloc, num_agents_accessible,
```
Should we use ptr_ here instead of ptr?  It might be a bit clearer to do so.

ampandey-AMD replied:
We want hsa_amd_pointer_info interception to be unwavered and return successfully when ASan is enabled. Here the user pointer `ptr` is not the actual ASan mapped `ptr_`. The ROCr runtime, given an interior pointer, returns the base of the whole mapped region (the page-aligned(kPageSize_) start: ptr_) when ASan is enabled. 
```
[guard/unused page] | [user payload of size m->UsedSize()]
^ ptr_ (GetBlockBegin)      ^ ptr (returned to user)
```
To keep the ASan illusion (that the user pointer is the “base” of the allocation and to hide the leading guard page), we shift the reported agentBaseAddress and hostBaseAddress forward by kPageSize_. We also replace sizeInBytes with `m->UsedSize()` (user payload only), instead of the full mapping size that includes the guard page.

If we passed ptr_ to the real function we would not add the offset, but then other interceptors (IPC create/attach) that rely on “user base = reported base” would break. 

The current chosen scheme below:

- Export full mapping when necessary (IPC create) using ptr_ and full length.
- For hsa_amd_pointer_info, keep calling with ptr so runtime finds the allocation, then adjust addresses to present the payload-only view.

b-sumner replied:
Is it not the case that the call at line 1562 to the real pointer info function will return the same result when given either ptr or ptr_?  If that is true then I think ptr_ is a better choice since it is the illusion free value passed to the illusion free function.

ampandey-AMD replied:
Ok, I tested with both ptr && ptr_ at line number 1562.

### Using ptr
```
$14 = {size = 56, type = HSA_EXT_POINTER_TYPE_HSA, agentBaseAddress = 0x7ffff435c000, hostBaseAddress = 0x7ffff435c000, sizeInBytes = 16384, userData = 0x0, agentOwner = {handle = 138331851461760},
  global_flags = 4, registered = true}
```

### Using ptr_
```
$16 = {size = 56, type = HSA_EXT_POINTER_TYPE_HSA, agentBaseAddress = 0x7ffff435c000, hostBaseAddress = 0x7ffff435c000, sizeInBytes = 16384, userData = 0x0, agentOwner = {handle = 138331851461760},
  global_flags = 4, registered = true}
```

Hi @bing-ma , should we replace ptr with ptr_ here at line 1562?

Also if we did replace it, should we also then remove the offset adjustment of agentBaseAddress, hostBaseAddress and sizeInBytes? 

I thought that we were trying to not alter the hsa_amd_pointer_info output in both asan/non-asan mode. Is that correct?

b-sumner replied:
Hi Amit.  I'm thinking that when we call the real function, we should use the real arguments.  But we need to adjust all caller inputs and outputs to maintain the illusion we are creating and that includes adjusting the pointer info output.

ampandey-AMD replied:
Ok, Done.

ampandey-AMD committed:
Address comments of @b-sumner.

Summary:

  - Switch `ptr` to `ptr_` when issuing real call to
    hsa_amd_pointer_info.
  - Add further comments to asan_hsa_amd_pointer_info implementation to
    depict adjustment of `info` o/p returned by hsa_amd_pointer_info.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2960

b-sumner reviewed (commented):
This looks good to me.  @bing-ma do you have any concerns?

bing-ma commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1540,5 +1547,34 @@ hsa_status_t asan_hsa_amd_vmem_address_free(void* ptr, size_t size,
+          reinterpret_cast<uptr>(info->hostBaseAddress) + kPageSize_);
+      info->sizeInBytes = m->UsedSize();
+    }
+  }
```
Should we add _else_ here to call the real hsa_amd_pointer_info() for non-ASAN allocations?

b-sumner replied:
That's a good question.  If we know nothing about the memory, the chances are good that HSA won't either, but maybe it would be better to fall back to the real function here like we do for IPC memory create.

ampandey-AMD replied:
Done.

bing-ma commented:
> This looks good to me. @bing-ma do you have any concerns?

Added a comment

ampandey-AMD committed:
Address comments.

  - Add fallback call to REAL hsa_amd_pointer_info with actual ptr.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2980

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2985

b-sumner commented on compiler-rt/lib/asan/asan_allocator.cpp:
```diff
@@ -1540,5 +1547,36 @@ hsa_status_t asan_hsa_amd_vmem_address_free(void* ptr, size_t size,
+  AsanChunk* m = ptr_
+                     ? instance.GetAsanChunkByAddr(reinterpret_cast<uptr>(ptr_))
+                     : nullptr;
+  hsa_status_t status = HSA_STATUS_ERROR_NOT_INITIALIZED;
```
We can eliminate the initialization, and only need "status" in the if block below if we return status from that block.  Then we can drop the else and unconditionally return the return from the HSA pointer info call after the if block.

ampandey-AMD committed:
Address comments of @b-sumner

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2995

ampandey-AMD committed:
Merge branch 'ROCm:amd-staging' into fix-swdev-563823-ipc-handles

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2997

b-sumner reviewed (commented):
Unless @bing-ma has additional comments, I think this is ready to go.

ampandey-AMD commented:
> Unless @bing-ma has additional comments, I think this is ready to go.

Ping

b-sumner reviewed (approved):
LGTM

bing-ma reviewed (approved):
LGTM
